opcua.cpp

```diff
@@ -285,7 +287,18 @@ DatapointValue* dpv = NULL;
 				dpname.erase(pos, 1);
 			}
 			points.push_back(new Datapoint(dpname, *dpv));
+
+                        SOPC_DateTime srcTimestamp = value->SourceTimestamp;
+                        time_t seconds;
+                        if (SOPC_Time_ToTimeT(srcTimestamp, &seconds) == SOPC_STATUS_OK)
+                        {
+                       		double TimeAsSecondsFloat = ((double) srcTimestamp) / 1.0E7;         // divide by 100 nanoseconds
+                                double integerPart = 0.0;
+                                m_userts.tv_sec = seconds;
+                                m_userts.tv_usec = (suseconds_t) (1E6 * modf(TimeAsSecondsFloat, &integerPart));
+                        }
 			ingest(points, 0);
```

Comment:
This is the only place where _OPCUA::ingest_ is called so it is easy to change its definition. It's second argument is currently "_long user_ts_" which is passed as zero so it is never used. In its implementation, it sets the Reading's user timestamp using the OPCUA class member _m_userts_. Instead, change the argument list of _OPCUA::ingest_ to _(vector<Datapoint *> points, timeval &user_ts)_. Create "_timeval userts_" as a local variable, do the math as you do now, and then pass _userts_ directly to _OPCUA::ingest_.

Reply:
fixed